types/index.d.ts

```diff
@@ -53,7 +53,7 @@ export class Engine {
 
   on(eventName: "success", handler: EventHandler): this;
   on(eventName: "failure", handler: EventHandler): this;
-  on(eventName: string, handler: EventHandler): this;
+  on<T>(eventName: string, handler: EventHandler<T>): this;
```

Comment:
@jagernet-ops you can simplify this whole block like this:
```ts
on<T = Event>(eventName: "success"|"failure", handler: EventHandler<T>): this;
```

Reply:
This update would force the value of `eventName` to be a string literal type of "success" or "failure." If users were to directly subscribe to a custom event then type errors will be thrown for the name of the event.